kedro-telemetry/kedro_telemetry/plugin.py

```diff
@@ -185,8 +193,10 @@ def _format_project_statistics_data(
 ):
     """Add project statistics to send to Heap."""
     project_statistics_properties = properties.copy()
-    project_statistics_properties["number_of_datasets"] = len(
-        catalog.datasets.__dict__.keys()
+    project_statistics_properties["number_of_datasets"] = sum(
+        1
+        for c in catalog.list()
+        if not c.startswith("parameters") and not c.startswith("params:")
```

Comment:
`catalog.list("^(?!params\:)\w+$")` will gives you the same thing I think. but I think this is also fine.

My question is should we also count the dataset generated from pattern? This will underestimate the datasets a lot if it's generated by patterns.

Reply:
@noklam , thank you for highlighting this issue. I've verified that our telemetry currently does not account for datasets created via dataset factories. 

While attempting to address this by scanning node.inputs and node.outputs, I encountered a challenge with intermediate datasets such as 'X_test', 'y_train', and 'y_test', which introduced noise. I also explored using catalog._datasets to identify Dataset types, but this approach doesn't reveal datasets instantiated by factories. 

Given these challenges, I propose we maintain the current state in this PR and open a separate issue dedicated to accurately counting datasets generated from dataset factories in telemetry.

Reply:
I agree with @DimedS 👍🏽 this is relevant for https://github.com/kedro-org/kedro/issues/1978 which is also in progress.